File: backend/src/presentation/api/routes/memories.py
"""メモリー記録管理API - UseCase Pattern"""

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from src.application.usecases.memory_record_usecase import MemoryRecordUseCase
from src.presentation.api.dependencies import get_memory_record_usecase

router = APIRouter(prefix="/memories", tags=["memories"])


class MemoryRecordCreateRequest(BaseModel):
    """メモリー記録作成リクエスト"""

    title: str
    description: str
    date: str
    type: str  # photo, video, album
    category: str  # milestone, daily, family, special
    media_url: str | None = None
    thumbnail_url: str | None = None
    location: str | None = None
    tags: list[str] = []
    favorited: bool = False
    user_id: str = "frontend_user"


class MemoryRecordUpdateRequest(BaseModel):
    """メモリー記録更新リクエスト"""

    title: str | None = None
    description: str | None = None
    date: str | None = None
    type: str | None = None
    category: str | None = None
    media_url: str | None = None
    thumbnail_url: str | None = None
    location: str | None = None
    tags: list[str] | None = None
    favorited: bool | None = None
    user_id: str = "frontend_user"


@router.post("/create")
async def create_memory_record(
    request: MemoryRecordCreateRequest,
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """メモリー記録を作成"""
    try:
        request_data = request.model_dump()
        user_id = request_data.get("user_id", "frontend_user")

        result = await memory_record_usecase.create_memory_record(user_id=user_id, record_data=request_data)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/list")
async def get_memory_records(
    user_id: str = "frontend_user",
    type: str | None = None,
    category: str | None = None,
    favorited: bool | None = None,
    tags: str | None = None,
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """メモリー記録一覧を取得"""
    try:
        filters = {}
        if type:
            filters["type"] = type
        if category:
            filters["category"] = category
        if favorited is not None:
            filters["favorited"] = favorited
        if tags:
            filters["tags"] = tags

        result = await memory_record_usecase.get_memory_records(user_id, filters)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/detail/{memory_id}")
async def get_memory_record(
    memory_id: str,
    user_id: str = "frontend_user",
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """特定のメモリー記録を取得"""
    try:
        result = await memory_record_usecase.get_memory_record(user_id, memory_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/update/{memory_id}")
async def update_memory_record(
    memory_id: str,
    request: MemoryRecordUpdateRequest,
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """メモリー記録を更新"""
    try:
        request_data = request.model_dump(exclude_unset=True)
        user_id = request_data.get("user_id", "frontend_user")

        result = await memory_record_usecase.update_memory_record(
            user_id=user_id, memory_id=memory_id, update_data=request_data,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/delete/{memory_id}")
async def delete_memory_record(
    memory_id: str,
    user_id: str = "frontend_user",
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """メモリー記録を削除"""
    try:
        result = await memory_record_usecase.delete_memory_record(user_id, memory_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.patch("/favorite/{memory_id}")
async def toggle_memory_favorite(
    memory_id: str,
    favorited: bool,
    user_id: str = "frontend_user",
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """メモリーのお気に入り状態を切り替え"""
    try:
        result = await memory_record_usecase.toggle_memory_favorite(
            user_id=user_id, memory_id=memory_id, favorited=favorited,
        )
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/favorites")
async def get_favorite_memories(
    user_id: str = "frontend_user",
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """お気に入りメモリーを取得"""
    try:
        result = await memory_record_usecase.get_favorite_memories(user_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/albums")
async def get_albums(
    user_id: str = "frontend_user",
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """アルバム一覧を取得"""
    try:
        result = await memory_record_usecase.get_albums(user_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/tags")
async def get_memory_tags(
    user_id: str = "frontend_user",
    memory_record_usecase: MemoryRecordUseCase = Depends(get_memory_record_usecase),
) -> dict[str, Any]:
    """使用中のタグ一覧を取得"""
    try:
        result = await memory_record_usecase.get_memory_tags(user_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/libs/api/memories.ts
/**
 * メモリー記録API関連の関数
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

export interface MemoryRecord {
  id: string
  user_id: string
  title: string
  description: string
  date: string
  type: 'photo' | 'video' | 'album'
  category: 'milestone' | 'daily' | 'family' | 'special'
  media_url?: string
  thumbnail_url?: string
  location?: string
  tags: string[]
  favorited: boolean
  created_at: string
  updated_at: string
}

export interface MemoryRecordCreateRequest {
  title: string
  description: string
  date: string
  type: 'photo' | 'video' | 'album'
  category: 'milestone' | 'daily' | 'family' | 'special'
  media_url?: string
  thumbnail_url?: string
  location?: string
  tags?: string[]
  favorited?: boolean
  user_id?: string
}

export interface MemoryRecordUpdateRequest {
  title?: string
  description?: string
  date?: string
  type?: 'photo' | 'video' | 'album'
  category?: 'milestone' | 'daily' | 'family' | 'special'
  media_url?: string
  thumbnail_url?: string
  location?: string
  tags?: string[]
  favorited?: boolean
  user_id?: string
}

export interface ApiResponse<T> {
  success: boolean
  data?: T
  message?: string
}

/**
 * メモリー一覧を取得
 */
export async function getMemories(params?: {
  user_id?: string
  type?: string
  category?: string
  favorited?: boolean
  tags?: string
}): Promise<ApiResponse<MemoryRecord[]>> {
  try {
    const searchParams = new URLSearchParams()
    
    if (params?.user_id) searchParams.append('user_id', params.user_id)
    if (params?.type) searchParams.append('type', params.type)
    if (params?.category) searchParams.append('category', params.category)
    if (params?.favorited !== undefined) searchParams.append('favorited', String(params.favorited))
    if (params?.tags) searchParams.append('tags', params.tags)

    const url = `${API_BASE_URL}/api/v1/memories/list${searchParams.toString() ? '?' + searchParams.toString() : ''}`
    
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('メモリー一覧取得エラー:', error)
    return {
      success: false,
      message: 'メモリー一覧の取得に失敗しました'
    }
  }
}

/**
 * メモリー詳細を取得
 */
export async function getMemory(memoryId: string, userId: string = 'frontend_user'): Promise<ApiResponse<MemoryRecord>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/memories/detail/${memoryId}?user_id=${userId}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('メモリー詳細取得エラー:', error)
    return {
      success: false,
      message: 'メモリー詳細の取得に失敗しました'
    }
  }
}

/**
 * 新しいメモリーを作成
 */
export async function createMemory(memoryData: MemoryRecordCreateRequest): Promise<ApiResponse<MemoryRecord>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/memories/create`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...memoryData,
        user_id: memoryData.user_id || 'frontend_user'
      }),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('メモリー作成エラー:', error)
    return {
      success: false,
      message: 'メモリーの作成に失敗しました'
    }
  }
}

/**
 * メモリーを更新
 */
export async function updateMemory(memoryId: string, memoryData: MemoryRecordUpdateRequest): Promise<ApiResponse<MemoryRecord>> {
  try {
    const response = await fetch(`${API_BASE_URL}/api/v1/memories/update/${memoryId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...memoryData,
        user_id: memoryData.user_id || 'frontend_user'
      }),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('メモリー更新エラー:', error)
    return {
      success: false,
      message: 'メモリーの更新に失敗しました'
    }
  }
}

/**
 * メモリーを削除
 */
export async function deleteMemory(memoryId: string, userId: string = 'frontend_user'): Promise<ApiResponse<MemoryRecord>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/memories/delete/${memoryId}?user_id=${userId}`,
      {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('メモリー削除エラー:', error)
    return {
      success: false,
      message: 'メモリーの削除に失敗しました'
    }
  }
}

/**
 * メモリーのお気に入り状態を切り替え
 */
export async function toggleMemoryFavorite(memoryId: string, favorited: boolean, userId: string = 'frontend_user'): Promise<ApiResponse<MemoryRecord>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/memories/favorite/${memoryId}?favorited=${favorited}&user_id=${userId}`,
      {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('お気に入り切り替えエラー:', error)
    return {
      success: false,
      message: 'お気に入りの切り替えに失敗しました'
    }
  }
}

/**
 * お気に入りメモリー一覧を取得
 */
export async function getFavoriteMemories(userId: string = 'frontend_user'): Promise<ApiResponse<MemoryRecord[]>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/memories/favorites?user_id=${userId}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('お気に入りメモリー取得エラー:', error)
    return {
      success: false,
      message: 'お気に入りメモリーの取得に失敗しました'
    }
  }
}

/**
 * アルバム一覧を取得
 */
export async function getAlbums(userId: string = 'frontend_user'): Promise<ApiResponse<MemoryRecord[]>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/memories/albums?user_id=${userId}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('アルバム取得エラー:', error)
    return {
      success: false,
      message: 'アルバムの取得に失敗しました'
    }
  }
}

/**
 * タグ一覧を取得
 */
export async function getMemoryTags(userId: string = 'frontend_user'): Promise<ApiResponse<string[]>> {
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/memories/tags?user_id=${userId}`,
      {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }
    )

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    return await response.json()
  } catch (error) {
    console.error('タグ取得エラー:', error)
    return {
      success: false,
      message: 'タグの取得に失敗しました'
    }
  }
}